# import Libraries

In [19]:
import pandas as pd
import numpy as np

# read dataset

In [20]:
books = pd.read_csv('book.csv',encoding='latin-1')
books.head()

Unnamed: 0  User.ID                                         Book.Title  \
0           1   276726                                Classical Mythology   
1           2   276729                                       Clara Callan   
2           3   276729                               Decision in Normandy   
3           4   276736  Flu: The Story of the Great Influenza Pandemic...   
4           5   276737                             The Mummies of Urumchi   

   Book.Rating  
0            5  
1            3  
2            6  
3            8  
4            6

# droping unwanted column

In [21]:
books.drop('Unnamed: 0',inplace=True,axis=1)
books.head()

User.ID                                         Book.Title  Book.Rating
0   276726                                Classical Mythology            5
1   276729                                       Clara Callan            3
2   276729                               Decision in Normandy            6
3   276736  Flu: The Story of the Great Influenza Pandemic...            8
4   276737                             The Mummies of Urumchi            6

# renaming column names

In [23]:
books=books.rename(columns={'User.ID':'ID', 'Book.Title':'Title', 'Book.Rating':'Rating'})
books.head()

ID                                              Title  Rating
0  276726                                Classical Mythology       5
1  276729                                       Clara Callan       3
2  276729                               Decision in Normandy       6
3  276736  Flu: The Story of the Great Influenza Pandemic...       8
4  276737                             The Mummies of Urumchi       6

# sorting id column from asending order

In [25]:
books=books.sort_values('ID')
books.head(10)

ID                                              Title  Rating
2401   8                                              Wings       5
2400   8  The Western way: A practical guide to the West...       5
2399   8                            Ancient Celtic Romances       5
2402   8                                           Truckers       5
2405   8                                  The Art Of Celtia       7
2403   8                Keepers of the Earth Teachers Guide       6
2404   8                            The Celts Activity Book       6
2406   9   The Book of Kells: Selected Plates in Full Color       6
2407  10                                SINGLE &amp; SINGLE       6
2408  12                                     Apocalypse Wow      10

from above we can say that user 8 have read 7 different books and his giving ratings is average above 5.

In [26]:
from pandas import Grouper
books.groupby(pd.Grouper(key="ID")).count()

Title  Rating
ID                   
8           7       7
9           1       1
10          1       1
12          1       1
14          3       3
...       ...     ...
278846      1       1
278849      1       1
278851     14      14
278852      1       1
278854      6       6

[2182 rows x 2 columns]

In [27]:
#number of unique users in the dataset
len(books.ID.unique())

2182

In [28]:
#count of rating column
books['Rating'].value_counts()

8     2283
7     2076
10    1732
9     1493
5     1007
6      920
4      237
3      146
2       63
1       43
Name: Rating, dtype: int64

maximum user gives 8 ratings

In [29]:
books.Title.unique()

array(['Wings',
       'The Western way: A practical guide to the Western mystery tradition',
       'Ancient Celtic Romances', ...,
       "Celtic Mythology (Library of the World's Myths and Legends)",
       'A corrente de Trewis Scott',
       'A Treasury of Irish Myth, Legend, and Folklore'], dtype=object)

In [31]:
#count of unique books
len(books.Title.unique())

9659

In [32]:
books.Title.value_counts().head(10)

Fahrenheit 451                                     5
The Subtle Knife (His Dark Materials, Book 2)      4
Stardust                                           4
Charlie and the Chocolate Factory                  4
Ender's Game (Ender Wiggins Saga (Paperback))      4
The Amber Spyglass (His Dark Materials, Book 3)    4
Vanished                                           4
Pet Sematary                                       3
Life Is Elsewhere                                  3
The Street Lawyer                                  3
Name: Title, dtype: int64

most read book is Fahrenheit 451

# deleting duplicated entries

In [37]:
books=books.drop_duplicates()

# duplicate entry of user with same book

In [38]:
books[books.duplicated(subset=["ID","Title"])]

ID                        Title  Rating
3786    1211                  Black House       8
4745    2033               The First Time       2
5218    2276  In the Beauty of the Lilies       9
9945  162052                         Rose       8
2001  278418     White Oleander : A Novel       5

In [41]:
bookclean=books.drop_duplicates(subset=["ID","Title"])
bookclean

ID                                              Title  Rating
2401       8                                              Wings       5
2400       8  The Western way: A practical guide to the West...       5
2399       8                            Ancient Celtic Romances       5
2402       8                                           Truckers       5
2405       8                                  The Art Of Celtia       7
...      ...                                                ...     ...
2396  278854                                Blast From the Past       7
2395  278854      La crónica del Perú (Crónicas de América)       7
2398  278854  Celtic Mythology (Library of the World's Myths...       8
2393  278854                         A corrente de Trewis Scott       7
2397  278854     A Treasury of Irish Myth, Legend, and Folklore       6

[9993 rows x 3 columns]

# sparse matrix

In [42]:
user_books = bookclean.pivot(index='ID',
                            columns='Title',
                            values='Rating')

In [43]:
user_books.head()

Title   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
ID                                                                        
8                       NaN                                         NaN   
9                       NaN                                         NaN   
10                      NaN                                         NaN   
12                      NaN                                         NaN   
14                      NaN                                         NaN   

Title   Repairing PC Drives &amp  '48  \
ID                                      
8                            NaN  NaN   
9                            NaN  NaN   
10                           NaN  NaN   
12                           NaN  NaN   
14                           NaN  NaN   

Title  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
ID                                                                             
8                                                    NaN                       
9                                                    NaN                       
10                                                   NaN                       
12                                                   NaN                       
14                                                   NaN                       

Title  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
ID                                                                    
8                                                    NaN              
9                                                    NaN              
10                                                   NaN              
12                                                   NaN              
14                                                   NaN              

Title  01-01-00: A Novel of the Millennium  \
ID                                           
8                                      NaN   
9                                      NaN   
10                                     NaN   
12                                     NaN   
14                                     NaN   

Title  1,401 More Things That P*Ss Me Off  10 Commandments Of Dating  \
ID                                                                     
8                                     NaN                        NaN   
9                                     NaN                        NaN   
10                                    NaN                        NaN   
12                                    NaN                        NaN   
14                                    NaN                        NaN   

Title  100 Great Fantasy Short, Short Stories  ...  \
ID                                             ...   
8                                         NaN  ...   
9                                         NaN  ...   
10                                        NaN  ...   
12                                        NaN  ...   
14                                        NaN  ...   

Title  Zora Hurston and the Chinaberry Tree (Reading Rainbow Book)  \
ID                                                                   
8                                                    NaN             
9                                                    NaN             
10                                                   NaN             
12                                                   NaN             
14                                                   NaN             

Title  \Even Monkeys Fall from Trees\ and Other Japanese Proverbs  \
ID                                                                  
8                                                    NaN            
9                                                    NaN            
10                                                   NaN            
12                                                   NaN            
14             

In [44]:
#impute those NaN with 0 values
user_books.fillna(0, inplace=True)

In [45]:
user_books

Title    Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
ID                                                                         
8                        0.0                                         0.0   
9                        0.0                                         0.0   
10                       0.0                                         0.0   
12                       0.0                                         0.0   
14                       0.0                                         0.0   
...                      ...                                         ...   
278846                   0.0                                         0.0   
278849                   0.0                                         0.0   
278851                   0.0                                         0.0   
278852                   0.0                                         0.0   
278854                   0.0                                         0.0   

Title    Repairing PC Drives &amp  '48  \
ID                                       
8                             0.0  0.0   
9                             0.0  0.0   
10                            0.0  0.0   
12                            0.0  0.0   
14                            0.0  0.0   
...                           ...  ...   
278846                        0.0  0.0   
278849                        0.0  0.0   
278851                        0.0  0.0   
278852                        0.0  0.0   
278854                        0.0  0.0   

Title   'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
ID                                                                              
8                                                     0.0                       
9                                                     0.0                       
10                                                    0.0                       
12                                                    0.0                       
14                                                    0.0                       
...                                                   ...                       
278846                                                0.0                       
278849                                                0.0                       
278851                                                0.0                       
278852                                                0.0                       
278854                                                0.0                       

Title   ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
ID                                                                     
8                                                     0.0              
9                                                     0.0              
10                                                    0.0              
12                                                    0.0              
14                                                    0.0              
...                                                   ...              
278846                                                0.0              
278849                                                0.0              
278851                                                0.0              
278852                                                0.0              
278854                                                0.0              

Title   01-01-00: A Novel of the Millennium  \
ID                                            
8                                       0.0   
9                                       0.0   
10                                      0.0   
12                                      0.0   
14                                      0.0   
...                                     ...   
278846                                  0.0   
278849                                  0.0   
278

# Calculating Cosine Similarity between Users

In [48]:
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine, correlation

In [51]:
user_sim = 1 - pairwise_distances( user_books.values, metric='cosine')

In [52]:
user_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [54]:
#Store the results in a dataframe
user_sim_df = pd.DataFrame(user_sim)

In [55]:
user_sim_df

0     1     2     3     4     5     6     7     8     9     ...  2172  \
0      1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1      0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2      0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
3      0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
4      0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
2177   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2178   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2179   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2180   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2181   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

      2173  2174  2175  2176  2177  2178  2179  2180  2181  
0      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
2177   0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0  
2178   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0  
2179   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0  
2180   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0  
2181   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0  

[2182 rows x 2182 columns]

In [57]:
#set the index and column names to user to user ids because above it automattically assign index no. not a user id 
user_sim_df.index = books.ID.unique()
user_sim_df.columns = books.ID.unique()

In [58]:
user_sim_df.iloc[0:10, 0:10]

8    9    10   12   14   16   17   19   22   26
8   1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
9   0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
10  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
12  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0
14  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0
16  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
17  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0
19  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
22  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
26  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0

In [59]:
 # Replace the 1 which is self make value replace by 0...not useful for further processs
np.fill_diagonal(user_sim, 0)
user_sim_df.iloc[0:10, 0:10]    

8    9    10   12   14   16   17   19   22   26
8   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
9   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
12  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
14  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
16  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
17  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
19  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
22  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
26  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0

In [60]:

user_sim_df.idxmax(axis=1)[0:5]

8     8
9     8
10    8
12    8
14    8
dtype: int64

In [61]:
bookclean[(bookclean['ID']==276729) |(bookclean['ID']==276726)]

ID                 Title  Rating
0  276726   Classical Mythology       5
2  276729  Decision in Normandy       6
1  276729          Clara Callan       3

### here 276729 and 276726 user have similiarties..so we can recommand 276729's watch movies to 276726 user.but highly recommmanded movie is 'Decision in Normandy' because it has high rating.